File: lib/models/event.dart
class Event {
  final String id;
  final String title;
  final String description;
  final DateTime startTime;
  final DateTime endTime;
  final List<String> participants;
  final String ownerId;
  final List<String> sharedWith;

  Event({
    required this.id,
    required this.title,
    required this.description,
    required this.startTime,
    required this.endTime,
    required this.participants,
    required this.ownerId,
    required this.sharedWith,
  });

  // Convert JSON to Event object
  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['_id'],
      title: json['title'],
      description: json['description'] ?? '',
      startTime: DateTime.parse(json['startTime']),
      endTime: DateTime.parse(json['endTime']),
      participants: List<String>.from(json['participants'] ?? []),
      ownerId: json['ownerId'],
      sharedWith: List<String>.from(json['sharedWith'] ?? []),
    );
  }

  // Convert Event object to JSON
  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'startTime': startTime.toIso8601String(),
      'endTime': endTime.toIso8601String(),
      'participants': participants,
      'ownerId': ownerId,
      'sharedWith': sharedWith,
    };
  }
}

File: server/models/event.js
import mongoose from 'mongoose';

const eventSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Event title is required'],
  },
  description: {
    type: String,
    default: '',
  },
  startTime: {
    type: Date,
    required: [true, 'Start time is required'],
  },
  endTime: {
    type: Date,
    required: [true, 'End time is required'],
  },
  participants: {
    type: [String], // Array of participant email strings
    default: [],
  },
  sharedWith: {
    type: [mongoose.Schema.Types.ObjectId], // Array of user IDs the event is shared with
    ref: 'User',
    default: [],
  },
  ownerId: {
    type: mongoose.Schema.Types.ObjectId, // Owner of the event
    ref: 'User',
    required: [true, 'Owner ID is required'],
  },
  comments: [
    {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'eventComment',
    },
  ],
}, { timestamps: true }); // Automatically adds `createdAt` and `updatedAt` fields

const Event = mongoose.model('Event', eventSchema);
export default Event;
